### PR TITLE
Use TemplateTypeRep in DAML Trigger API

### LINKED ISSUE #3215: Use TemplateTypeRep in DAML Trigger API
This allows us to make ACS lookups more efficient and it allows us to have typed contract ids.

### PR DESCRIPTION
**Pull request by aherrmann-da**
Closes #3215 

- Add a template-type indexed version of `AnyContractId`.
    - Adds type-safety when handling contract Ids.
    - Avoids type parameter on `exerciseCmd`.
    - Easier matching on `ArchivedEvent`, e.g. `fromArchived` instead of `cid.templateId.entityName == "XYZ"`.
- Group contracts in `ACS` by `TemplateTypeRep`.
    - Allows more efficient lookups, e.g. in `getTemplates`.
    - Requires `MapKey` instance for `TemplateTypeRep`.
- Adds `toString` and `fromString` on `Identifier`.
    Uses format `<packageId>:<moduleName>:<entityName>`.
- Replaces `Identifier` by `TemplateTypeRep` in trigger runner.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: triggers/daml/Daml/Trigger/LowLevel.daml [hunk 1]
```diff
@@ -26,17 +30,27 @@ module Daml.Trigger.LowLevel
...
-
 data AnyContractId = AnyContractId
-  { templateId : Identifier
+  { templateId : TemplateTypeRep
   , contractId : Text
   } deriving (Show, Eq, Ord)
 
+-- XXX: The name ContractId clashes with Prelude.ContractId.
+--   Find a better name than TheContractId.
+newtype TheContractId t = TheContractId { contractId : Text }
```
Comment (cocreature) [inline triggers/daml/Daml/Trigger/LowLevel.daml]: Hm, not quite sure if having to hide Prelude is better or worse. Slightly leaning towards just requiring hiding Prelude and/or qualified imports but neither is great.

Maybe we can come up with some terrible hack to get to and from text functions for absolute contract ids and then just use the actual ContractId type but I’m not quite sure how to pull that off.

Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: It's tricky, requiring to hide `Prelude.ContractId` increases the boilerplate for creating triggers. Also, trigger modules might still require `Prelude.ContractId`, e.g. [here](https://github.com/digital-asset/daml/blob/6c36777a8ee65e0942233d3bea02bc29b4e583d9/triggers/tests/daml/ACS.daml#L91), so using the same name can be confusing.

Comment (cocreature) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: Maybe `AbsoluteContractId` isn’t so bad? The normal `ContractId` can be both relative (while building up a transaction) and absolute which is the part that stops us from implementing conversions to `Text` so this at least conveys the right distinction for why we can do things here that we can’t do with the normal `ContractId` type.

Comment (aherrmann-da) [reply triggers/daml/Daml/Trigger/LowLevel.daml]: Yes, that sounds better. I've changed it to that for now.

**Pull request merged**